<!-- markdownlint-disable -->
# no-invalid-hint-params

## Description

Checks if hints are parameterized correctly

## Metadata

- Fixable: ❌
- Suggestions: ❌
- Recommended: ✅
- Type: problem

## Options

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "array",
  "items": [],
  "minItems": 0
}
```

### Default options

```json
[]
```

## Rule source

https://github.com/AdguardTeam/AGLint/src/rules/no-invalid-hint-params.ts

## Test cases

https://github.com/AdguardTeam/AGLint/test/rules/no-invalid-hint-params.test.ts

