<!-- markdownlint-disable -->
# no-invalid-domains

## Description

Disallows invalid domains

## Metadata

- Fixable: ❌
- Suggestions: ✅
- Recommended: ✅
- Type: problem

## Options

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "array",
  "items": [],
  "minItems": 0
}
```

### Default options

```json
[]
```

## Rule source

https://github.com/AdguardTeam/AGLint/src/rules/no-invalid-domains.ts

## Test cases

https://github.com/AdguardTeam/AGLint/test/rules/no-invalid-domains.test.ts

